### ISSUE TITLE
Ingress for CAPO MC

### ISSUE THREAD

**Issue by cornelius-keller** [giantswarm/roadmap#583]

**Comment by kopiczko** [comment]
There is a problem that cloud-provider creates LB with the same name for every cluster.

<img width="1422" alt="Screenshot 2021-11-25 at 11 15 16" src="https://user-images.githubusercontent.com/4608994/143432497-6010ff08-a578-4459-8179-067422be3b9f.png">

```
$ kubectl get service -n kube-system nginx-ingress-controller --context gravity-admin@gravity
NAME                       TYPE           CLUSTER-IP     EXTERNAL-IP       PORT(S)                      AGE
nginx-ingress-controller   LoadBalancer   10.102.22.35   216.119.153.220   80:30326/TCP,443:31194/TCP   43h
```

```
$ kubectl get service -n kube-system nginx-ingress-controller --context galaxy-admin@galaxy
NAME                       TYPE           CLUSTER-IP       EXTERNAL-IP       PORT(S)                      AGE
nginx-ingress-controller   LoadBalancer   10.109.160.118   216.119.153.181   80:31035/TCP,443:31725/TCP   47h
```

Basted on the IP seems like this LB belongs to `gravity` and I don't see a LB for `galaxy`.

When curling the connection doesn't seem very reliable even for the `.220` (for `.181` it doesn't work at all):

```
$ for i in $(seq 1 5); do echo "--------------------"; curl -m 2 http://216.119.153.220 ; done
--------------------
curl: (28) Operation timed out after 2006 milliseconds with 0 bytes received
--------------------
curl: (28) Operation timed out after 2004 milliseconds with 0 bytes received
--------------------
<html>
<head><title>404 Not Found</title></head>
<body>
<center><h1>404 Not Found</h1></center>
<hr><center>nginx</center>
</body>
</html>
--------------------
<html>
<head><title>404 Not Found</title></head>
<body>
<center><h1>404 Not Found</h1></center>
<hr><center>nginx</center>
</body>
</html>
--------------------
curl: (28) Operation timed out after 2005 milliseconds with 0 bytes received
```

```
$ for i in $(seq 1 5); do echo "--------------------"; curl -m 2 http://216.119.153.181 ; done
--------------------
curl: (28) Connection timed out after 2005 milliseconds
--------------------
curl: (28) Connection timed out after 2004 milliseconds
--------------------
curl: (28) Connection timed out after 2001 milliseconds
--------------------
curl: (28) Connection timed out after 2003 milliseconds
--------------------
curl: (28) Connection timed out after 2000 milliseconds
```

From `gravity`:

```
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager I1123 16:04:15.491884       1 controllermanager.go:275] Starting "cloud-node"
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager I1123 16:04:15.496098       1 node_controller.go:115] Sending events to api server.
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager I1123 16:04:15.496192       1 controllermanager.go:285] Started "cloud-node"
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager I1123 16:04:15.496200       1 controllermanager.go:275] Starting "cloud-node-lifecycle"
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager I1123 16:04:15.496278       1 node_controller.go:154] Waiting for informer caches to sync
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager I1123 16:04:15.498383       1 node_lifecycle_controller.go:76] Sending events to api server
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager I1123 16:04:15.498442       1 controllermanager.go:285] Started "cloud-node-lifecycle"
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager I1123 16:04:15.498450       1 controllermanager.go:275] Starting "service"
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager I1123 16:04:15.504009       1 openstack.go:335] Claiming to support LoadBalancer
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager I1123 16:04:15.504027       1 controllermanager.go:285] Started "service"
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager I1123 16:04:15.504035       1 controllermanager.go:275] Starting "route"
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager I1123 16:04:15.504110       1 controller.go:230] Starting service controller
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager I1123 16:04:15.504122       1 shared_informer.go:240] Waiting for caches to sync for service
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager W1123 16:04:15.551212       1 openstack.go:444] Error initialising Routes support: router-id not set in cloud provider config
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager W1123 16:04:15.551242       1 core.go:110] --configure-cloud-routes is set, but cloud provider does not support routes. Will not configure cloud provider routes.
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager W1123 16:04:15.551248       1 controllermanager.go:282] Skipping "route"
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager I1123 16:04:15.597473       1 node_controller.go:390] Initializing node gravity-control-plane-dfpvp with cloud provider
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager I1123 16:04:15.604173       1 shared_informer.go:247] Caches are synced for service
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager I1123 16:04:16.250098       1 openstack.go:342] Claiming to support Zones
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager I1123 16:04:16.411472       1 node_controller.go:454] Successfully initialized node gravity-control-plane-dfpvp with cloud provider
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager I1123 16:04:16.411533       1 node_controller.go:390] Initializing node gravity-md-0-cs7wq with cloud provider
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager I1123 16:04:16.411677       1 event.go:291] "Event occurred" object="gravity-control-plane-dfpvp" kind="Node" apiVersion="v1" type="Normal" reason="Synced" message="Node synced successfully"
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager I1123 16:04:17.050019       1 openstack.go:342] Claiming to support Zones
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager I1123 16:04:17.208843       1 node_controller.go:454] Successfully initialized node gravity-md-0-cs7wq with cloud provider
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager I1123 16:04:17.208877       1 node_controller.go:390] Initializing node gravity-md-0-6gf6j with cloud provider
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager I1123 16:04:17.209120       1 event.go:291] "Event occurred" object="gravity-md-0-cs7wq" kind="Node" apiVersion="v1" type="Normal" reason="Synced" message="Node synced successfully"
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager I1123 16:04:17.850480       1 openstack.go:342] Claiming to support Zones
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager I1123 16:04:18.005768       1 node_controller.go:454] Successfully initialized node gravity-md-0-6gf6j with cloud provider
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager I1123 16:04:18.005823       1 node_controller.go:390] Initializing node gravity-md-0-6dtkg with cloud provider
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager I1123 16:04:18.005940       1 event.go:291] "Event occurred" object="gravity-md-0-6gf6j" kind="Node" apiVersion="v1" type="Normal" reason="Synced" message="Node synced successfully"
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager I1123 16:04:18.809083       1 openstack.go:342] Claiming to support Zones
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager I1123 16:04:18.951087       1 node_controller.go:454] Successfully initialized node gravity-md-0-6dtkg with cloud provider
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager I1123 16:04:18.951163       1 event.go:291] "Event occurred" object="gravity-md-0-6dtkg" kind="Node" apiVersion="v1" type="Normal" reason="Synced" message="Node synced successfully"
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager I1123 16:17:33.133420       1 event.go:291] "Event occurred" object="kube-system/nginx-ingress-controller" kind="Service" apiVersion="v1" type="Normal" reason="EnsuringLoadBalancer" message="Ensuring load balancer"
openstack-cloud-controller-manager-rbspt openstack-cloud-controller-manager I1123 16:18:29.465032       1 event.go:291] "Event occurred" object="kube-system/nginx-ingress-controller" kind="Service" apiVersion="v1" type="Normal" reason="EnsuredLoadBalancer" message="Ensured load balancer"
```

From `galaxy`:

```
openstack-cloud-controller-manager-59m6r openstack-cloud-controller-manager E1123 13:18:20.080214       1 controller.go:307] error processing service kube-system/nginx-ingress-controller (will retry): failed to ensure load balancer: error creating LB floatingip: Resource not found: {"NeutronError": {"type": "ExternalGatewayForFloatingIPNotFound", "message": "External network 7135bb8d-95e1-4e81-b183-83ccd1354a82 is not reachable from subnet a89fd22b-098c-4a4f-8212-b5b613ab7ba3.  Therefore, cannot associate Port d21cca46-43bd-49a7-af39-6ce1a644e7d3 with a Floating IP.", "detail": ""}}
openstack-cloud-controller-manager-59m6r openstack-cloud-controller-manager I1123 13:18:20.080413       1 event.go:291] "Event occurred" object="kube-system/nginx-ingress-controller" kind="Service" apiVersion="v1" type="Warning" reason="SyncLoadBalancerFailed" message="Error syncing load balancer: failed to ensure load balancer: error creating LB floatingip: Resource not found: {\"NeutronError\": {\"type\": \"ExternalGatewayForFloatingIPNotFound\", \"message\": \"External network 7135bb8d-95e1-4e81-b183-83ccd1354a82 is not reachable from subnet a89fd22b-098c-4a4f-8212-b5b613ab7ba3.  Therefore, cannot associate Port d21cca46-43bd-49a7-af39-6ce1a644e7d3 with a Floating IP.\", \"detail\": \"\"}}"
openstack-cloud-controller-manager-59m6r openstack-cloud-controller-manager I1123 13:23:20.082550       1 event.go:291] "Event occurred" object="kube-system/nginx-ingress-controller" kind="Service" apiVersion="v1" type="Normal" reason="EnsuringLoadBalancer" message="Ensuring load balancer"
openstack-cloud-controller-manager-59m6r openstack-cloud-controller-manager E1123 13:23:24.620100       1 controller.go:307] error processing service kube-system/nginx-ingress-controller (will retry): failed to ensure load balancer: error creating LB floatingip: Resource not found: {"NeutronError": {"type": "ExternalGatewayForFloatingIPNotFound", "message": "External network 7135bb8d-95e1-4e81-b183-83ccd1354a82 is not reachable from subnet a89fd22b-098c-4a4f-8212-b5b613ab7ba3.  Therefore, cannot associate Port d21cca46-43bd-49a7-af39-6ce1a644e7d3 with a Floating IP.", "detail": ""}}
openstack-cloud-controller-manager-59m6r openstack-cloud-controller-manager I1123 13:23:24.620372       1 event.go:291] "Event occurred" object="kube-system/nginx-ingress-controller" kind="Service" apiVersion="v1" type="Warning" reason="SyncLoadBalancerFailed" message="Error syncing load balancer: failed to ensure load balancer: error creating LB floatingip: Resource not found: {\"NeutronError\": {\"type\": \"ExternalGatewayForFloatingIPNotFound\", \"message\": \"External network 7135bb8d-95e1-4e81-b183-83ccd1354a82 is not reachable from subnet a89fd22b-098c-4a4f-8212-b5b613ab7ba3.  Therefore, cannot associate Port d21cca46-43bd-49a7-af39-6ce1a644e7d3 with a Floating IP.\", \"detail\": \"\"}}"
openstack-cloud-controller-manager-59m6r openstack-cloud-controller-manager I1123 13:28:24.621511       1 event.go:291] "Event occurred" object="kube-system/nginx-ingress-controller" kind="Service" apiVersion="v1" type="Normal" reason="EnsuringLoadBalancer" message="Ensuring load balancer"
openstack-cloud-controller-manager-59m6r openstack-cloud-controller-manager E1123 13:28:29.224723       1 controller.go:307] error processing service kube-system/nginx-ingress-controller (will retry): failed to ensure load balancer: error creating LB floatingip: Resource not found: {"NeutronError": {"type": "ExternalGatewayForFloatingIPNotFound", "message": "External network 7135bb8d-95e1-4e81-b183-83ccd1354a82 is not reachable from subnet a89fd22b-098c-4a4f-8212-b5b613ab7ba3.  Therefore, cannot associate Port d21cca46-43bd-49a7-af39-6ce1a644e7d3 with a Floating IP.", "detail": ""}}
openstack-cloud-controller-manager-59m6r openstack-cloud-controller-manager I1123 13:28:29.224956       1 event.go:291] "Event occurred" object="kube-system/nginx-ingress-controller" kind="Service" apiVersion="v1" type="Warning" reason="SyncLoadBalancerFailed" message="Error syncing load balancer: failed to ensure load balancer: error creating LB floatingip: Resource not found: {\"NeutronError\": {\"type\": \"ExternalGatewayForFloatingIPNotFound\", \"message\": \"External network 7135bb8d-95e1-4e81-b183-83ccd1354a82 is not reachable from subnet a89fd22b-098c-4a4f-8212-b5b613ab7ba3.  Therefore, cannot associate Port d21cca46-43bd-49a7-af39-6ce1a644e7d3 with a Floating IP.\", \"detail\": \"\"}}"
openstack-cloud-controller-manager-59m6r openstack-cloud-controller-manager I1123 13:33:29.226838       1 event.go:291] "Event occurred" object="kube-system/nginx-ingress-controller" kind="Service" apiVersion="v1" type="Normal" reason="EnsuringLoadBalancer" message="Ensuring load balancer"
openstack-cloud-controller-manager-59m6r openstack-cloud-controller-manager E1123 13:33:31.000392       1 controller.go:307] error processing service kube-system/nginx-ingress-controller (will retry): failed to ensure load balancer: error creating LB floatingip: Resource not found: {"NeutronError": {"type": "ExternalGatewayForFloatingIPNotFound", "message": "External network 7135bb8d-95e1-4e81-b183-83ccd1354a82 is not reachable from subnet a89fd22b-098c-4a4f-8212-b5b613ab7ba3.  Therefore, cannot associate Port d21cca46-43bd-49a7-af39-6ce1a644e7d3 with a Floating IP.", "detail": ""}}
openstack-cloud-controller-manager-59m6r openstack-cloud-controller-manager I1123 13:33:31.000729       1 event.go:291] "Event occurred" object="kube-system/nginx-ingress-controller" kind="Service" apiVersion="v1" type="Warning" reason="SyncLoadBalancerFailed" message="Error syncing load balancer: failed to ensure load balancer: error creating LB floatingip: Resource not found: {\"NeutronError\": {\"type\": \"ExternalGatewayForFloatingIPNotFound\", \"message\": \"External network 7135bb8d-95e1-4e81-b183-83ccd1354a82 is not reachable from subnet a89fd22b-098c-4a4f-8212-b5b613ab7ba3.  Therefore, cannot associate Port d21cca46-43bd-49a7-af39-6ce1a644e7d3 with a Floating IP.\", \"detail\": \"\"}}"
openstack-cloud-controller-manager-59m6r openstack-cloud-controller-manager I1123 13:38:31.002276       1 event.go:291] "Event occurred" object="kube-system/nginx-ingress-controller" kind="Service" apiVersion="v1" type="Normal" reason="EnsuringLoadBalancer" message="Ensuring load balancer"
openstack-cloud-controller-manager-59m6r openstack-cloud-controller-manager I1123 13:38:40.898123       1 event.go:291] "Event occurred" object="kube-system/nginx-ingress-controller" kind="Service" apiVersion="v1" type="Normal" reason="EnsuredLoadBalancer" message="Ensured load balancer"
```

Note that in both cases there is `"Event occurred" object="kube-system/nginx-ingress-controller" kind="Service" apiVersion="v1" type="Normal" reason="EnsuredLoadBalancer" message="Ensured load balancer"` but `galaxy` also outputs some errors like `"Event occurred" object="kube-system/nginx-ingress-controller" kind="Service" apiVersion="v1" type="Warning" reason="SyncLoadBalancerFailed" message="Error syncing load balancer: failed to ensure load balancer: error creating LB floatingip: Resource not found: {\"NeutronError\": {\"type\": \"ExternalGatewayForFloatingIPNotFound\", \"message\": \"External network 7135bb8d-95e1-4e81-b183-83ccd1354a82 is not reachable from subnet a89fd22b-098c-4a4f-8212-b5b613ab7ba3.  Therefore, cannot associate Port d21cca46-43bd-49a7-af39-6ce1a644e7d3 with a Floating IP.\", \"detail\": \"\"}}"`.

**Comment by kopiczko** [comment]
We figured so far that creating a LB with the same name is possible, but not sure why cloud-controller is not doing this.

Aslo it looks like cloud-controller takes down the LB it doesn't own if LB Service is deleted (looks like it's matching by name).

**Comment by kopiczko** [comment]
There is a problem with being able to have only one ingress LB per cluster: https://gigantic.slack.com/archives/C01BYMF6RN0/p1638296207459800

This is because LB is referenced by a name in the cloud controller (see the Slack thread linked above).

**Comment by kopiczko** [comment]
This is mostly done. There are some cleanup PRs hanging:

- https://github.com/giantswarm/nginx-ingress-controller-app/pull/256 - to avoid some confusion.
- https://github.com/kubernetes/cloud-provider-openstack/pull/1717 - upstream, this is the thing we struggled to configure with @ericgraf. If the change is accepted next time we'll be at least warned with a log.

Besides that I opened https://github.com/giantswarm/roadmap/issues/625 to restore `externalTrafficPolicy: Local` but it isn't crucial right now.
